src/rust/jsg/wrappable.rs

```diff
@@ -0,0 +1,222 @@
+        unsafe { v8::ffi::unwrap_string(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+impl Type for bool {
+    type This = Self;
+
+    fn class_name() -> &'static str {
+        "boolean"
+    }
+
+    fn wrap<'a, 'b>(this: Self::This, lock: &'a mut Lock) -> v8::Local<'b, v8::Value>
+    where
+        'b: 'a,
+    {
+        this.to_local(lock)
+    }
+
+    fn is_exact(value: &v8::Local<v8::Value>) -> bool {
+        value.is_boolean()
+    }
+
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self {
+        unsafe { v8::ffi::unwrap_boolean(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+impl Type for f64 {
+    type This = Self;
+
+    fn class_name() -> &'static str {
+        "number"
+    }
+
+    fn wrap<'a, 'b>(this: Self::This, lock: &'a mut Lock) -> v8::Local<'b, v8::Value>
+    where
+        'b: 'a,
+    {
+        this.to_local(lock)
+    }
+
+    fn is_exact(value: &v8::Local<v8::Value>) -> bool {
+        value.is_number()
+    }
+
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self {
+        unsafe { v8::ffi::unwrap_number(isolate.as_ffi(), value.into_ffi()) }
+    }
+}
+
+/// Trait for converting between Rust and JavaScript values.
+///
+/// Provides bidirectional conversion: `wrap` converts Rust to JavaScript,
+/// `unwrap` converts JavaScript to Rust. The `try_unwrap` method is used
+/// by macros to unwrap parameters with proper error handling.
+pub trait Wrappable: Sized {
+    /// Converts this Rust value into a JavaScript value.
+    fn wrap(self, lock: &mut Lock) -> v8::Local<'_, v8::Value>;
+
+    /// Converts a JavaScript value into this Rust type.
+    fn unwrap(isolate: v8::IsolatePtr, value: v8::Local<v8::Value>) -> Self;
+
+    /// Attempts to unwrap a JavaScript value, returning None and throwing a JS error on failure.
+    /// Most types simply delegate to `unwrap`, but `NonCoercible<T>` validates the type first.
+    fn try_unwrap(lock: &mut Lock, value: v8::Local<v8::Value>) -> Option<Self> {
+        Some(Self::unwrap(lock.isolate(), value))
+    }
+
+    /// Wraps this value and sets it as the function return value.
+    fn wrap_return(self, lock: &mut Lock, args: &mut v8::FunctionCallbackInfo) {
+        args.set_return_value(self.wrap(lock));
+    }
+}
+
+impl Wrappable for () {
+    fn wrap(self, lock: &mut Lock) -> v8::Local<'_, v8::Value> {
+        v8::Local::<v8::Value>::undefined(lock)
+    }
+
+    fn unwrap(_isolate: v8::IsolatePtr, _value: v8::Local<v8::Value>) -> Self {}
+}
+
+impl<T: Wrappable, E: Display> Wrappable for Result<T, E> {
+    fn wrap(self, lock: &mut Lock) -> v8::Local<'_, v8::Value> {
+        match self {
+            Ok(value) => value.wrap(lock),
+            Err(_) => unreachable!("Cannot wrap Result::Err"),
+        }
+    }
+
+    fn unwrap(_isolate: v8::IsolatePtr, _value: v8::Local<v8::Value>) -> Self {
+        unreachable!("Cannot unwrap into Result")
```

Comment:
here too: if this is unreachable then unwrap should not be part of Wrappable but be in Unwrappable. Otherwise someone will crash the code.